### PR TITLE
Updates for supporting sample exports

### PR DESCRIPTION
**Pull request by rahul-tuli**
This PR adds support to export sample inputs and outputs for SparseML's YOLOv5 integration

### CODE REVIEW

File: utils/sparse.py [hunk 1]
```diff
@@ -158,3 +162,56 @@ def _apply_one_shot(self):
...
+
+        sample_in_dir = os.path.join(save_dir, "sample-inputs")
+        sample_out_dir = os.path.join(save_dir, "sample-outputs")
+
+        os.makedirs(sample_in_dir, exist_ok=True)
+        os.makedirs(sample_out_dir, exist_ok=True)
+
+        for _, (images, _, _, _) in enumerate(dataloader):
+            images = (
+                images.float() / 255
```
Comment (KSGulin) [inline utils/sparse.py:190]: To be thorough, we may want to add multi_scaling here if it was turned on for the run. See lines 176 - 182 in #40

**Review by KSGulin — COMMENTED** [review]
Great addition, very clean code 🙂

**Review by KSGulin — APPROVED** [review]

File: utils/sparse.py (hunk 1 above)
Comment (bfineran) [inline utils/sparse.py:190]: does the dataloader give INT8 inputs? shouldn't it already include preprocessing?

File: utils/sparse.py [hunk 2]
```diff
@@ -158,3 +162,56 @@ def _apply_one_shot(self):
...
+            ins = images.detach().to("cpu")
+            outs = [elem.detach().to("cpu") for elem in outs]
+            outs_gen = zip(*outs)
+
+            for sample_in, sample_out in zip(ins, outs_gen):
+                sample_out = list(sample_out)
+                file_idx = f"{exported_samples}".zfill(4)
+
+                numpy.savez(f"{sample_in_dir}/inp-{file_idx}.npz", sample_in)
+                numpy.savez(f"{sample_out_dir}/out-{file_idx}.npz", *sample_out)
```
Comment (bfineran) [inline utils/sparse.py:206]: should use `os.path.join` `/` can change between systems

Comment (bfineran) [inline utils/sparse.py:206]: are these being saved as dictionaries?

File: utils/sparse.py (hunk 1 above)
Comment (rahul-tuli) [reply utils/sparse.py:190]: Yes the dataloader gives INT8 inputs, the inference code is essentially the same as in ultralytics yolo train script

File: utils/sparse.py (hunk 2 above)
Comment (rahul-tuli) [reply utils/sparse.py:206]: Agree! good catch, addressed!

Comment (rahul-tuli) [reply utils/sparse.py:206]: No, YOLO returns a list of 3 items as output for each image, that's why the `*`

Comment (bfineran) [reply utils/sparse.py]: Good point. Default behavior on ONNX export is to name the outputs `output_0`...`output_{n}`.  Let's make it a dict with those names to match ONNX

**Review by bfineran — CHANGES_REQUESTED** [review]
left update to comment about assigning default names for output tensors. LGTM after!

File: utils/sparse.py (hunk 2 above)
Comment (rahul-tuli) [reply utils/sparse.py]: I checked the Sample Outputs from SparseZoo and they follow the same default numpy names

**Review by bfineran — APPROVED** [review]

**Pull request merged**